Release notes v0.36 — quantum algorithms section (#5574)
PennyLaneAI/pennylane: doc/releases/changelog-0.36.0.md

<h4>Access an extended arsenal of quantum algorithms 🏹</h4>

> * The Fast Approximate BLock-Encodings (FABLE) algorithm for embedding

Author: And added the qubitization entry!

  a matrix into a quantum circuit

>   (https://arxiv.org/abs/2205.00081) is now accessible via the `qml.FABLE` 

Author: ```suggestion
  [arXiv:2205.00081](https://arxiv.org/abs/2205.00081) is now accessible via the `qml.FABLE` 
```

  template.
  [(#5107)](https://github.com/PennyLaneAI/pennylane/pull/5107)

  The usage of `qml.FABLE` is similar to `qml.BlockEncode` but provides a more
  efficient circuit construction at the cost of a user-defined approximation 
  level, `tol`. The number of wires that `qml.FABLE` operates on is `2*n + 1`, 
  where `n` defines the dimension of the :math:`2^n \times 2^n` matrix that we
  want to block-encode.

  ```python
  A = np.array([[0.1, 0.2], [0.3, 0.4]])
  dev = qml.device('default.qubit')

  @qml.qnode(dev)
  def circuit():
      qml.FABLE(A, tol=0.1, wires=range(3))
      return qml.state()
  ``` 

  ```pycon
  >>> circuit()
  tensor([0.04929927+0.j, 0.15067163+0.j, 0.04929927+0.j, 0.15067163+0.j,
          0.49756365+0.j, 0.47675786+0.j, 0.49756365+0.j, 0.47675786+0.j], requires_grad=True)
  ```

> * Reflecting about a given quantum state is now available via `qml.Reflection`.

Author: @trbromley re: https://github.com/PennyLaneAI/pennylane/pull/5574#discussion_r1584879331

I thought you were referencing the other comment you made about `Reflection` being put above `AmplitudeAmplification` 😅. If I read the changelog top-down, `Reflection` being put first makes sense to me because it builds into / is a subroutine in `AmplitudeAmplification`

Reviewer: I dropped a comment in that thread too, but my thinking is that it's nicer to see the important thing first, i.e., amplitude amplification. Then, by the way, we have support for `Reflection` too which is an important component of that.

But this doesn't block merging the PR though.

Reviewer 2: I agree with Tom  :) `AmplitudeAmplification` is the important one here

    decorate the quantum function with `@qml.prod`, which turns the quantum 
    function into a `Prod` instance:

    ```python
    @qml.prod
    def U():
        qml.Hadamard(wires=0)
        qml.RY(0.1, wires=1)

    @qml.qnode(dev)
    def circuit():
        qml.Reflection(U())
        return qml.state()

    circuit()
    ```

    ```pycon
    >>> circuit()
    tensor([-0.00249792-6.13852933e-17j,  0.04991671+3.05651685e-18j,
          0.99750208+6.10793866e-17j,  0.04991671+3.05651685e-18j], requires_grad=True)
    ```
  
* A high-level interface for amplitude amplification and its variants is now 
  available via the new `qml.AmplitudeAmplification` template.
  [(#5160)](https://github.com/PennyLaneAI/pennylane/pull/5160)

  Based on [arXiv:quant-ph/0005055](https://arxiv.org/abs/quant-ph/0005055), 
  given a state :math:`\vert \Psi \rangle = \alpha \vert \phi \rangle + \beta \vert \phi^{\perp} \rangle`, 
  `qml.AmplitudeAmplification` amplifies the amplitude of :math:`\vert \phi \rangle`. 
  Mathematically, 
  :math:`\texttt{AmplitudeAmplification} \vert \Psi \rangle \approx \vert \phi \rangle`.

  `qml.AmplitudeAmplification` requires

>   * `U`: an `Operator` that prepares :math:`\vert \Psi \rangle` (i.e., 

Author: Am trying a couple things to see if this gets fixed. No idea why this is happening 🤔

>   * `U`: an `Operator` that prepares :math:`\vert \Psi \rangle` (i.e., 

Author: POURQUOI

![Screenshot 2024-04-30 at 5 03 39 PM](https://github.com/PennyLaneAI/pennylane/assets/34751083/1d494f45-598a-490d-b40c-6c02ca840336)

>   * `U`: an `Operator` that prepares :math:`\vert \Psi \rangle` (i.e., 

Author: oh there's a typo 😆. Let's see what the fix brings...

>   * `U`: an `Operator` that prepares :math:`\vert \Psi \rangle` (i.e., 

Reviewer: Still not showing up nicely for me :thinking: I don't have a suggestion on how to fix this though. We could just remove the "`qml.AmplitudeAmplification` requires" section, since they can look in the docstrings.

  level, `tol`. The number of wires that `qml.FABLE` operates on is `2*n + 1`, 
  where `n` defines the dimension of the :math:`2^n \times 2^n` matrix that we
  want to block-encode.

  ```python
  A = np.array([[0.1, 0.2], [0.3, 0.4]])
  dev = qml.device('default.qubit', wires=3)

  @qml.qnode(dev)
  def circuit():
      qml.FABLE(A, tol = 0.001, wires=range(3))  
      return qml.state()
  ``` 

  ```pycon
  >>> mat = qml.matrix(circuit)()
  >>> 2 * mat[0:2, 0:2]
  array([[0.1+0.j, 0.2+0.j],
         [0.3+0.j, 0.4+0.j]])
  ```

* A high-level interface for amplitude amplification and its variants is now 
  available via the new `qml.AmplitudeAmplification` template.
  [(#5160)](https://github.com/PennyLaneAI/pennylane/pull/5160)

  Based on [arXiv:quant-ph/0005055](https://arxiv.org/abs/quant-ph/0005055), 
  given a state :math:`\vert \Psi \rangle = \alpha \vert \phi \rangle + \beta \vert \phi^{\perp} \rangle`, 
  `qml.AmplitudeAmplification` amplifies the amplitude of :math:`\vert \phi \rangle`. 
  Mathematically, 
  :math:`\texttt{AmplitudeAmplification} \vert \Psi \rangle \approx \vert \phi \rangle`.

>   Here's an example with 

Reviewer: ```suggestion
  Here's an example with a target state
```

Reviewer: Just adding these suggestions because we took at the added information about `U` and `O`.

  :math:`\vert \phi \rangle = \vert 2 \rangle = \vert 010 \rangle`,

>   :math:`U = H^{\otimes 3}`, and an operator `O` being modelled as a `qml.FlipSign` 
>   operation (it is a required argument that flips the sign of 
>   :math:`\vert \phi \rangle` and does nothing to 
>   :math:`\vert \phi^{\perp} \rangle`):

Reviewer: ```suggestion
  an input state :math:`\vert \Psi \rangle = H^{\otimes 3} \vert 000 \rangle`, as well as an
  oracle that flips the sign of :math:`\vert \phi \rangle` and does nothing to
  :math:`\vert \phi^{\perp} \rangle`, which can be achieved in this case through
  `qml.FlipSign`.
```

  @qml.qnode(dev)
  def circuit():

>       generator(wires=range(3)) # prepares |Psi> = U|0>

Reviewer: Suggestion for line 88:
````md
  ```

  Here `U` is a quantum operation that is created by decorating a quantum function
  with `@qml.prod`. This could alternatively be done by creating a user-defined
  [custom operation](https://docs.pennylane.ai/en/stable/development/adding_operators.html)
  with a decomposition. Amplitude amplification can then be set up within a circuit:

  ```python
````

      qml.AmplitudeAmplification(U, O, iters=10)

      return qml.probs(wires=range(3))
  ```
  
  ```pycon
  >>> print(np.round(circuit(), 3))
  [0.01  0.01  0.931 0.01  0.01  0.01  0.01  0.01 ]
  ```

  As expected, we amplify the :math:`\vert 2 \rangle` state.

* Reflecting about a given quantum state is now available via `qml.Reflection`.

>   This operation is very useful in, say, the amplitude amplification algorithm.

Reviewer: ```suggestion
  This operation is very useful in the amplitude amplification algorithm and offers a generalization
  of `qml.FlipSign` which operates on basis states.
```

  [(#5159)](https://github.com/PennyLaneAI/pennylane/pull/5159)

  `qml.Reflection` works by providing an operation, :math:`U`, that *prepares* the 
  desired state, :math:`\vert \psi \rangle`, that we want to reflect about. In other 
  words, :math:`U` is such that :math:`U \vert 0 \rangle = \vert \psi \rangle`. In 
  PennyLane, :math:`U` must be an `Operator`.
  
  For example, if we want to reflect about 
  :math:`\vert \psi \rangle = \vert + \rangle`, then :math:`U = H`:

  ```python
  U = qml.Hadamard(wires=wires)

  dev = qml.device('default.qubit')
  @qml.qnode(dev)
  def circuit():
        qml.Reflection(U)
        return qml.state()
  ```

  ```pycon
  >>> circuit()
  tensor([0.-6.123234e-17j, 1.+6.123234e-17j], requires_grad=True)
  ```

>   For cases where :math:`U` comprises many operations, you can formulate it as an 
>   `Operator` by 
>   
>   * creating a user-defined 
>     [custom operation](https://docs.pennylane.ai/en/stable/development/adding_operators.html) 
>     with a decomposition, or by
> 
>   * creating a quantum function containing each operation, one per line, then 
>     decorate the quantum function with `@qml.prod`, which turns the quantum 
>     function into a `Prod` instance:
> 
>     ```python
>     @qml.prod
>     def U():
>         qml.Hadamard(wires=0)
>         qml.RY(0.1, wires=1)
> 
>     @qml.qnode(dev)
>     def circuit():
>         qml.Reflection(U())
>         return qml.state()
>     ```
> 

Reviewer: ```suggestion
```
Suggest to move some of this content further up

  @qml.qnode(dev)
  def circuit():
      generator(wires=range(3)) # prepares |Psi> = U|0>
      qml.AmplitudeAmplification(U, O, iters=10)

      return qml.probs(wires=range(3))
  ```
  
  ```pycon
  >>> print(np.round(circuit(), 3))
  [0.01  0.01  0.931 0.01  0.01  0.01  0.01  0.01 ]
  ```

  As expected, we amplify the :math:`\vert 2 \rangle` state.

* Performing qubitization is now easily accessible with the new 
  `qml.Qubitization` operator.
  [(#5500)](https://github.com/PennyLaneAI/pennylane/pull/5500)

  `qml.Qubitization` encodes a Hamiltonian into a suitable unitary operator. 
  When applied in conjunction with quantum phase estimation (QPE), it allows 
  for computing the eigenvalue of an eigenvector of the given Hamiltonian. 

  ```python
  H = qml.dot([0.1, 0.3, -0.3], [qml.Z(0), qml.Z(1), qml.Z(0) @ qml.Z(2)])
  @qml.qnode(qml.device("default.qubit"))
  def circuit():
      # initialize the eigenvector
      qml.PauliX(2)
      # apply QPE
      measurements = qml.iterative_qpe(
          qml.Qubitization(H, control = [3,4]), ancilla = 5, iters = 3
      )
      return qml.probs(op = measurements)
  

>   output = circuit()

Author: I think so? Although I think it should be `-0.7` since that's the ground state. But since there's non-negligible confusion here we can take it out.
